### PR TITLE
GH-2252: Keep offset metadata in case of batch reprocessing

### LINKED ISSUE #2252: Offset metadata should not be lost in case of batch reprocessing
Issue #2170 introduced a way to control the metadata associated to an offset but in case of exception during the processing of a batch, these metadata are no present anymore when records are reprocessed.

So, in this issue, we will ensure these metadata are not lost in such a case.

Javadoc of OffsetAndMetadataProvider has also been reviewed as this provider is used for both, sync and async commits.

### PR DESCRIPTION
**Pull request by frosiere**
Remarks:

- Every OffsetAndMetadata are now created using the provider if available. 
- Listener info has been added to the MessageListener container interface.
- Code related to the provider has been refactored and simplified.
- Javadoc has been reviewed.
- Updated the CommitProperties documentation

Closes #2252

### CODE REVIEW

**Review by artembilan — CHANGES_REQUESTED** [review]
Please, add your name to the `@author` list of all the affected classes.
When it missed yet, of course.

Otherwise this is OK with me even if those minor concerns.

Thank you!

**Review by garyrussell — CHANGES_REQUESTED** [review]
Just one minor nit.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultListenerMetadata.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * Default implementation for {@link ListenerMetadata}.
+ * @author Francois Rosiere
+ * @since 2.8.6
+ */
+public class DefaultListenerMetadata implements ListenerMetadata {
+
+	private final MessageListenerContainer container;
+
+	public DefaultListenerMetadata(MessageListenerContainer container) {
+		this.container = container;
```
[the remark this answers is not in the document]
Comment (frosiere) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultListenerMetadata.java:33]: Check has been added.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultListenerMetadata.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+package org.springframework.kafka.listener;
+
+import org.springframework.lang.Nullable;
+
+/**
+ * Default implementation for {@link ListenerMetadata}.
+ * @author Francois Rosiere
+ * @since 2.8.6
+ */
+public class DefaultListenerMetadata implements ListenerMetadata {
```
[the remark this answers is not in the document]
Comment (frosiere) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultListenerMetadata.java]: Removed the public modifier.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/MessageListenerContainer.java [hunk 3]
```diff
@@ -194,6 +195,17 @@ default String getListenerId() {
 		throw new UnsupportedOperationException("This container does not support retrieving the listener id");
 	}
 
+	/**
+	 * Get arbitrary static information that will be added to the
+	 * {@link KafkaHeaders#LISTENER_INFO} header of all records.
+	 * @return the info.
+	 * @since 2.8.4
```
[the remark this answers is not in the document]
Comment (frosiere) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/MessageListenerContainer.java]: Updated the version.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultAfterRollbackProcessor.java [hunk 4]
```diff
@@ -165,4 +167,10 @@ public void clearThreadState() {
 		this.lastIntervals.remove();
 	}
 
+	private OffsetAndMetadata createOffsetAndMetadata(@Nullable MessageListenerContainer container, long offset) {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultAfterRollbackProcessor.java:170]: `static`

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultAfterRollbackProcessor.java (hunk 4 above)
Comment (frosiere) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultAfterRollbackProcessor.java]: Missing static keyword added.

**Review by artembilan — APPROVED** [review]

**Comment by frosiere** [comment]
Any review remarks have been applied. If something else is required, please let me know.

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
Many thanks for the contribution.

**Pull request merged**